src/Microsoft.ML.FastTree/Representation/TreeRegressor.cs

```diff
@@ -0,0 +1,169 @@
+﻿using System;
+using System.Linq;
+using Microsoft.ML.Trainers.FastTree.Internal;
+
+namespace Microsoft.ML.FastTree.Representation
+{
+    /// <summary>
+    /// A container class for exposing <see cref="RegressionTree"/>'s attributes to users.
+    /// This class should not be mutable, so it contains a lot of read-only members.
+    /// </summary>
+    public class TreeRegressor
+    {
+        /// <summary>
+        /// <see cref="TreeRegressor"/> is an immutable wrapper over <see cref="_tree"/> for exposing some tree's
+        /// attribute to users.
+        /// </summary>
+        private readonly RegressionTree _tree;
+
+        /// <summary>
+        /// Sample labels from training data. <see cref="_leafSamples"/>[i] stores the labels falling into the
+        /// i-th leaf.
+        /// </summary>
+        private readonly double[][] _leafSamples;
+        /// <summary>
+        /// Sample labels' weights from training data. <see cref="_leafSampleWeights"/>[i] stores the weights for
+        /// labels falling into the i-th leaf. <see cref="_leafSampleWeights"/>[i][j] is the weight of
+        /// <see cref="_leafSamples"/>[i][j].
+        /// </summary>
+        private readonly double[][] _leafSampleWeights;
```

Comment:
So is this an attempt to reflect the polymorphism present in what is currently the internal class `RegressionTree`?

Reply:
Let's not fake polymorphism. I think you ought to have an abstract class, possibly, `RegressionTreeBase`. It works over regression tree. You have two subclasses: `RegressionTree` (whose implementation is almost entirely empty), and `QuantileRegressionTree`, the latter working over quantile regression trees. Then your `TreeEnsemble` is generic over the type of tree, which must descend from `RegressionTreeBase`. E.g.: `public sealed class TreeEnsemble<TTree> where TTree : RegressionTreeBase`. Then the specific model parameter objects have as something accessible either `TreeEnsemble<RegressionTree>` or `TreeEnsemble<QuantileRegressionTree>`.

Note that I am using the desirable names for these things, not this "regressor" or "collection" stuff.

---
In reply to: [252500060](https://github.com/dotnet/machinelearning/pull/2243#discussion_r252500060) [](ancestors = 252500060)

Reply:
Changes for addressing this comment were proposed in [this commit](https://github.com/dotnet/machinelearning/pull/2243/commits/9ffc6c8a9e1304e3554a34e1bcd70b11da00cb11). Could you please take a look?